ENS subgraph incremental PR 1

rahul7668gupta opened the pull request:
hosted service deployment can be found here: https://thegraph.com/hosted-service/subgraph/rahul7668gupta/ens-subgraph
Note: this deployment keeps on getting redeployed as it is in development phase.

deepesh-kn commented on airstack-modules/bin/index.ts:
```diff
@@ -117,12 +117,164 @@ integrate(vertical, yaml, graphql, dataSources, templates)
+          tokenAddress: string,
+        )
+
+        For documentation and examples please check: https://github.com/Airstack-xyz/airstack-subgraph/tree/task/npx-scripts
```
This should point to the main branch

rahul7668gupta replied:
done

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+    let domainId = createAirDomainEntityId(node, label);
+    let domain = getOrCreateAirDomain(new Domain(domainId, chainId, block, tokenAddress));
+    if (fromOldRegistry && domain.isMigrated == true) {
+      // this domain was migrated from the old registry, so we don't need to hanlde old registry event now
```
Please specify the reason. 
As a reviewer I am not sure on why

If we are ignoring the events form old registry, then check this at the first place at line 59

rahul7668gupta replied:
also, its being checked first thing

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+      tokenAddress,
+    ));
+    if (domain.parent == null && parent != null) {
+      parent.subdomainCount = parent.subdomainCount.plus(BIGINT_ONE);
```
domain.parent is `null`
This means that the domain is not the subdomain, then why we are incrementing the count?
Or please connect me on the call, as I am not clear with this condition.

rahul7668gupta replied:
we are not incrementing the count for domain.parent, a parent entity is being created and if domain already doesn't have a parent `if(domain.parent == null && parent != null)`, the parent entity subdomain count is increased.
Then, this parent entity is being assigned as `domain.parent=parent.id` below.

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+    blockHash: string,
+    blockTimestamp: BigInt,
+    logIndex: BigInt,
+    transactionHash: Bytes,
```
```suggestion
    transactionHash: string,
```

rahul7668gupta replied:
done

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+   * @param tokenAddress contract address of nft token
+   * @returns parent domain id
+   */
+  function recurseDomainDelete(domain: AirDomain, chainId: string, block: AirBlock, tokenAddress: string): string | null {
```
You are updating the subdomainCount, can we rename the function name? `recurseDomainDelete` sound like the entities are being deleted.

rahul7668gupta replied:
done

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+   * @param block air block object
+   * @returns updated entity count
+   */
+  function updateAirEntityCounter(
```
Move it to a common file.

rahul7668gupta replied:
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate entity counters.

rahul7668gupta replied:
but, they can do it without these functions also, as they will have full access to these entities eitherways, so, I am making this change.
DONE

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+   * @param slug subgraph slug
+   * @param name subgraph name
+   */
+  function createAirMeta(
```
Move to a common file.

rahul7668gupta replied:
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate air meta

rahul7668gupta replied:
but, they can do it without these functions also, as they will have full access to these entities eitherways, so, I am making this change.
DONE

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+   * @param blockTimestamp block timestamp
+   * @returns AirBlock entity
+   */
+  function getOrCreateAirBlock(
```
Move to a common file.

rahul7668gupta replied:
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate air block.

rahul7668gupta replied:
but, they can do it without these functions also, as they will have full access to these entities eitherways, so, I am making this change.
DONE

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+   * @param address token address
+   * @returns AirToken entity
+   */
+  function getOrCreateAirToken(chainID: string, address: string): AirToken {
```
Move to a common file.

rahul7668gupta replied:
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate air token

rahul7668gupta replied:
also, nfs-marketplace and domain-name verticals have a different schema for air token

rahul7668gupta replied:
cannot do, coz air token in domain-name vertical only has id and address fields, while nft-marketplace vertical has a lot of other fields also

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+   * @param address account address
+   * @returns AirAccount entity
+   */
+  function getOrCreateAirAccount(chainId: string, address: string): AirAccount {
```
Move to a common file.

rahul7668gupta replied:
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate air account

rahul7668gupta replied:
also, nft-marketplace and domain-name verticals have a different schema for air account

rahul7668gupta replied:
cannot do, coz air account in domain-name vertical only has id and address fields, while nft-marketplace vertical has a lot of other fields also

0xsarvesh replied:
Both should have same schema.

deepesh-kn commented on airstack-modules/modules/airstack/domain-name/utils.ts:
```diff
@@ -0,0 +1,76 @@
+AIR_NETWORK_MAP.set("matic", "MATIC");
+AIR_NETWORK_MAP.set("xdai", "XDAI");
+
+export function processNetwork(network: string): string {
```
In this file, please move the code to common util files so that other verticals can utilize it

rahul7668gupta replied:
done

deepesh-kn commented on airstack-modules/package.json:
```diff
@@ -4,13 +4,11 @@
     "test": "echo \"Error: no test specified\" && exit 1",
     "publish": "npm publish --tag alpha --access public",
-    "build": "npm run codegen && npm run build:subgraph && npm run build:package && npm pack"
+    "build": "npm run build:package && npm pack"
```
Locally, we will get errors for the missing generated files, 
Please revert this code. 
That was the reason we have graphql and yaml files in the project.

rahul7668gupta replied:
Not sure why this is still needed, I have built the module successfully multiple times without this.
Also, without graphql and yaml files, vs code shows errors for imports from ../generated/schema, but when imported inside the subgraph code via npx, it works great.

rahul7668gupta replied:
I can understand it doesn't look good in vscode problems section, but I think that is an insufficient reason to keep redundant files inside the codebase. I think it would just confuse devs for the existence of those files.

deepesh-kn reviewed (dismissed):
I only went through the module code.
Providing the first set of comments, I am still going through the logical things.

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+    logIndex: BigInt,
+    chainId: string,
+    newOwner: string,
+    transactionHash: Bytes,
```
done

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+    newOwner: string,
+    transactionHash: Bytes,
+    isMigrated: boolean,
+    node: Bytes,
```
node and label from bytes to string cannot be done, as crypto.keccak256 function takes both inputs as a byte to create id, otherwise it wont work

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1074 @@
+    transactionHash: Bytes,
+    isMigrated: boolean,
+    node: Bytes,
+    label: Bytes,
```
node and label from bytes to string cannot be done, as crypto.keccak256 function takes both inputs as a byte to create id, otherwise it wont work

rahul7668gupta commented on airstack-modules/modules/airstack/domain.ts:
```diff
@@ -0,0 +1,98 @@
+
+  }
+
+  export function updateAirEntityCounter(
```
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate entity counters.

rahul7668gupta replied:
but, they can do it without these functions also, as they will have full access to these entities eitherways, so, I am making this change.

rahul7668gupta replied:
done

rahul7668gupta commented on airstack-modules/modules/airstack/domain.ts:
```diff
@@ -0,0 +1,98 @@
+    return entity.count;
+  }
+
+  export function createAirMeta(
```
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate air meta

rahul7668gupta replied:
but, they can do it without these functions also, as they will have full access to these entities eitherways, so, I am making this change.
DONE

rahul7668gupta commented:
@deepesh-kn , I have addressed your comments, added my reasons for not making some changes and made changes for required ones, some in this same PR, some in another PR #36 . I have also asked @0xsarvesh to review this new PR.

0xsarvesh reviewed (changes requested):
Looking at the implementation.  Module code is very specific to ENS,  not sure if can be extended to other similar protocols like unstoppable domains

0xsarvesh commented on ens/tests/.bin/ens-registry.wat:
```diff
@@ -0,0 +1,12068 @@
+  (type (;7;) (func (param i32 i32 i32 i32) (result i32)))
+  (type (;8;) (func (param i32 i32 i32 i32)))
+  (type (;9;) (func (param i32)))
+  (type (;10;) (func (param i32 i32 i32 i32 i32 i32 i32)))
```
delete the .bin folder & add to gitignore.

rahul7668gupta replied:
done

rahul7668gupta commented on ens/src/reverse-registrar.ts:
```diff
@@ -0,0 +1,24 @@
+    call.inputs.name,
+    ETHEREUM_MAINNET_ID,
+    call.outputs.value0,
+    call.from.toHexString(),
```
I see the code treats from address as resolved address

rahul7668gupta replied:
initially I was treating the set name call from owner perspective, but now its being treated from resolved address perspective

rahul7668gupta commented on ens/src/ens-registry.ts:
```diff
@@ -0,0 +1,175 @@
+    event.params.node,
+    event.params.label,
+    TOKEN_ADDRESS_ENS,
+    false,
```
part of module refactor

0xsarvesh replied:
I still feel this can be covered in this PR. You can place below code in calling code, outside module. 
```
if (fromOldRegistry && isMigratedMapping.isMigrated == true) {
      // this domain was migrated from the old registry, so we don't need to hanlde old registry event now
      return;
    }
   ```

rahul7668gupta replied:
done

rahul7668gupta commented on ens/src/ens-registry.ts:
```diff
@@ -0,0 +1,175 @@
+    ETHEREUM_MAINNET_ID,
+    event.params.owner.toHexString(),
+    event.transaction.hash.toHexString(),
+    true,
```
part of module refactor

0xsarvesh replied:
Same point, lets move this outside module.

rahul7668gupta replied:
done

rahul7668gupta commented on ens/src/eth-registrar.ts:
```diff
@@ -0,0 +1,162 @@
+  log.info("handleNameRegistered: registrant {} label {} expiry {} txhash {}", [event.params.owner.toHexString(), event.params.id.toHexString(), event.params.expires.toString(), event.transaction.hash.toHexString()]);
+  let paymentToken: string | null = null
+  if (event.transaction.value > BIG_INT_ZERO) {
+    paymentToken = ZERO_ADDRESS;
```
done

rahul7668gupta commented on ens/src/eth-registrar.ts:
```diff
@@ -0,0 +1,162 @@
+    ETHEREUM_MAINNET_ID,
+    event.params.owner.toHexString(),
+    event.params.expires,
+    event.transaction.value,
```
linked event is coming from old registrar controller address which is being tracked with `handleNameRegisteredByController` and `handleNameRegisteredByControllerOld` event handlers.

This line is handling the event coming from Base Registrar which doesn't have cost. That's why the cost is being taken from the txn.value. We can remove the cost from this event if required.

0xsarvesh replied:
Make cost( transaction value) optional.  Other event is providing correct cost.

rahul7668gupta replied:
done in #39

rahul7668gupta commented on ens/src/eth-registrar.ts:
```diff
@@ -0,0 +1,162 @@
+    event.block.hash.toHexString(),
+    event.block.timestamp,
+    ETHEREUM_MAINNET_ID,
+    null,
```
this is a name renewed event coming from Base registrar contract which does not provide any cost related information.

name renewed event from eth registrar controller provides cost param and is being handled by `handleNameRenewedByController`

0xsarvesh replied:
what action is done for this?

rahul7668gupta replied:
creates air name renewed txn and updates label name and name in domain entity

rahul7668gupta commented on ens/src/eth-registrar.ts:
```diff
@@ -0,0 +1,162 @@
+  if (event.params.cost > BIG_INT_ZERO) {
+    paymentToken = ZERO_ADDRESS;
+  }
+  airstack.domain.trackSetNamePreImage(
```
part of module refactor

0xsarvesh replied:
I think rename of function should be done.

rahul7668gupta replied:
renaming to trackNameRenewedOrRegistrationByController

rahul7668gupta replied:
done

0xsarvesh replied:
Controller is specific to ENS

rahul7668gupta replied:
yes, I thought of moving this function outside module, but then we Will also have to expose get or create funcs for air block. I have updated the readme and function docs to specify that this function is specific to ens subgraph.

rahul7668gupta commented on ens/schema.graphql:
```diff
@@ -1,197 +1,184 @@
+  owner: AirAccount!
+  resolver: AirResolver
+  ttl: BigInt
+  isMigrated: Boolean!
```
remove is migrated

rahul7668gupta replied:
part of module refactor

rahul7668gupta replied:
done

rahul7668gupta replied:
created a mapping entity for this

rahul7668gupta commented on ens/src/eth-registrar.ts:
```diff
@@ -0,0 +1,162 @@
+    rootNode,
+    TOKEN_ADDRESS_ENS,
+    event.transaction.hash.toHexString(),
+    null,
```
this is a name registered event handler so, it doesn't have `renewer`. As for the expiry, it is already being tracked from the `handleNameRegistered` handler for brass registrar

rahul7668gupta commented on ens/src/resolver.ts:
```diff
@@ -0,0 +1,44 @@
+ */
+export function handleVersionChanged(event: VersionChangedEvent): void {
+  log.info("handleVersionChanged: node {} resolver {} txhash {}", [event.params.node.toHexString(), event.address.toHexString(), event.transaction.hash.toHexString()]);
+  airstack.domain.trackResolverVersionChange(
```
I think this can happen with any domain subgraph, the handler basically tracks the event when an owner changes the underlying resolver for a domain name. That sets the default resolved address for a domain as null.

0xsarvesh replied:
Lets confirm on this.

rahul7668gupta replied:
I did some digging into unstoppable domains and was unable to find out any version changed events as per docs/ resolver/ registry contracts.

0xsarvesh replied:
Let's do this at the time of module refactoring.

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+    chainId: string,
+    newOwner: string,
+    transactionHash: string,
+    isMigrated: boolean,
```
part of module refactor

0xsarvesh replied:
This should be removed.

rahul7668gupta replied:
done

rahul7668gupta replied:
done in next PR

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+    let block = getOrCreateAirBlock(chainId, blockHeight, blockHash, blockTimestamp);
+    let domainId = createAirDomainEntityId(node, label);
+    let domain = getOrCreateAirDomain(new Domain(domainId, chainId, block, tokenAddress));
+    if (fromOldRegistry && domain.isMigrated == true) {
```
no, the code checks if the domain linked to the event is already migrated to new contract or not. If it is and the event if from old registry, we don't need to process the event.
Otherwise it sets the isMigrated field as true or false depending on if the event is from new or old contract.

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+    fromOldRegistry: boolean,
+  ): void {
+    let block = getOrCreateAirBlock(chainId, blockHeight, blockHash, blockTimestamp);
+    let domain = getOrCreateAirDomain(new Domain(node, chainId, block, tokenAddress));
```
the domain entity has some required fields which have to be set with a default value when creating.
and the data to update this entity is scattered in a lot of different events. that's why the domain entity updation code is being written after a default domain entity has been created, otherwise the subgraph sync will keep on failing.

rahul7668gupta replied:
updates to the domain entity are done inside the track funcs

0xsarvesh replied:
@rahul7668gupta  To fix this.

rahul7668gupta replied:
updated owner for a domain in #39

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+    domain.resolver = resolverEntity.id;
+    // update domain resolved address
+    if (resolverEntity.resolvedAddress) {
+      domain.resolvedAddress = resolverEntity.resolvedAddress;
```
yes, they are different, but, after a resolved address is changed for a specific domain, it updates resolver.resolvedAddress from AddrChanged event.

Here, we are simply checking if that value exists (meaning an addr changed event already come through and updated the resolvedAddress for domain entity) in the resolver entity and assigning it to the domain.resolvedAddress field.

Also, this resolver entity is specific to a domain and the resolved Address too, as this entity is created with `resolver.concat("-").concat(node)`

rahul7668gupta replied:
code is correct

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+    }
+    // do recursive domain delete
+    domain.lastBlock = block.id;
+    recurseSubdomainCountDecrement(domain, chainId, block, tokenAddress);
```
this is because when you change the resolver for a domain, your subdomains are no longer linked to the same resolver.
But, I would like to investigate more on this.

rahul7668gupta replied:
Found this issue from the ens subgraph GitHub repo - https://github.com/ensdomains/ens-subgraph/pull/44

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+    blockTimestamp: BigInt,
+    logIndex: BigInt,
+    chainId: string,
+    registrantAddress: string,
```
that is because owner changes twice in while registering a domain, once its from 0x to contract and then from contract to actual owner. That is why this is being handled in the domain owner changed handler.

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+  }
+
+  /**
+   * @dev This function tracks set name preimage transaction
```
can we sync on this and understand the kind of information we would like to expose from this doc? 
I added the function description and signatures similar to the nft model for now.

rahul7668gupta replied:
need to sync on this

rahul7668gupta replied:
update the Bin and readme files

rahul7668gupta commented on ens/src/eth-registrar.ts:
```diff
@@ -0,0 +1,162 @@
+ * @dev this function maps the NameRegistered event from the EthRegistrarController contract
+ * @param event ControllerNameRegisteredEvent from EthRegistrarController contract
+ */
+export function handleNameRegisteredByController(event: ControllerNameRegisteredEvent): void {
```
that is because owner changes twice in while registering a domain, once its from 0x to contract and then from contract to actual owner. That is why this is being handled in the domain owner changed handler.

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+    }
+    if (domain.labelName !== name) {
+      domain.labelName = name
+      domain.name = name + '.eth'
```
part of module refactor

rahul7668gupta replied:
moved outside the code

rahul7668gupta replied:
done

rahul7668gupta commented on airstack-modules/modules/airstack/domain-name/domain-name.ts:
```diff
@@ -0,0 +1,1152 @@
+    resolver.domain = domain.id;
+    resolver.save()
+
+    if (domain.resolver == resolver.id) {
```
this is because people can set the default resolved address for a domain to any address, even if the address is not owned by them.
this check ensures that the domain resolver and resolver emitted in the resolvedAddress (Addr) changed event is same.

rahul7668gupta commented:
I have made logical changes and deployed the subgraph instance here - https://thegraph.com/hosted-service/subgraph/rahul7668gupta/ens-subgraph-1?version=pending

rahul7668gupta commented:
I have made logical changes and deployed the subgraph instance here - https://thegraph.com/hosted-service/subgraph/rahul7668gupta/ens-subgraph-1?version=pending

rahul7668gupta commented:
@0xsarvesh , please don't resolve conversations in this PR. Just approve if you think the logical changes are done.

rahul7668gupta commented on ens/src/ens-registry.ts:
```diff
@@ -0,0 +1,175 @@
+export function handleNewOwner(event: NewOwnerEvent): void {
+  log.info("handleNewOwner: owner {} node {} label {} txhash {}", [event.params.owner.toHexString(), event.params.node.toHexString(), event.params.label.toHexString(), event.transaction.hash.toHexString()]);
+  airstack.domain.trackDomainOwnerChangedTransaction(
+    event.block.number,
```
done

rahul7668gupta commented:
@0xsarvesh , I have addressed the discussed changes and pushed to #39

0xsarvesh reviewed (approved):
LGTM
